### DISCUSSION TITLE
Exposing client side ES modules

### DISCUSSION THREAD

**Discussion by nantunes** [vercel/next.js#40681] [Help]
I'm trying to share UI functionality between two Next.js apps, using ES modules.

I'm moving away from non-SSR apps using the SystemJS module format and wanting to avoid webpack's module federation scheme, hoping that the current standard ESM support across browsers and server is enough.

[Next.js 11.1 blog post](https://nextjs.org/blog/next-11-1#es-modules-support) mentions work "on extensive ES Modules support in Next.js, **both** as input modules and as an **output target**" (emphasis added). However [Next.js 12 post](https://nextjs.org/blog/next-12#es-modules-support-and-url-imports) seems to focus only on importing ES modules.

I've tried using a custom webpack config, enabling `config.experiments.outputModule`, but it only affected the server side code (generated files in `.next/server/pages` did become ESM - breaking the build as expected with syntax errors). Using `isServer` to configure only the client side code seems to have no effect (code in `.next/static` stays the same).

Is this unsupported or am I doing something wrong? Thanks!

**Edit:**
Reposted in stack overflow: https://stackoverflow.com/questions/73829389/next-js-exposing-client-side-es-modules

**Comment by jeremydrichardson** [comment]
I've been following this as well and here is my conclusions so far:

**Next JS always depends on build** - the architecture of Next JS is dependent on all modules being available at build time and doesn't look like that is changing anytime soon.

**URL imports** - this is an experimental feature that allow you to import directly from URLs. However this only happens during the build (from what I can tell) so you can't dynamically swap out the module at that url at runtime.

**Module federation** - still the only way I can find to provide module at runtime. According the @ScriptedAlchemy, Vercel has stated it will never support module federation (though I think that's only through his sources, I can't find any 3rd party verification of this).

I'd love to know if you're finding anything different.
